routing/index_graph.hpp

```diff
@@ -76,13 +207,171 @@ class IndexGraph final
+
+  /// \brief Fills |connectionPaths| with all path from joint |from| to joint |to|.
+  /// If |from| and |to| don't belong to the same feature |connectionPaths| will
+  /// be empty.
+  /// If |from| and |to| belong to only one feature |connectionPaths| will have one item.
+  /// It's most common case.
+  /// If |from| and |to| could be connected by several features |connectionPaths|
+  /// will have several items.
+  /// \note The order on points in items of |connectionPaths| is from |from| to |to|.
+  void GetConnectionPaths(Joint::Id from, Joint::Id to,
+                          vector<vector<RoadPoint>> & connectionPaths);
+
+  void CreateFakeFeatureGeometry(vector<RoadPoint> const & geometrySource, double speed,
+                                 RoadGeometry & geometry);
+
+  void GetNeighboringEdge(RoadGeometry const & road, RoadPoint const & rp, bool forward, bool outgoing,
+                          bool graphWithoutRestrictions, vector<JointEdge> & edges) const;
+
+  double GetSpeed(RoadPoint const & rp);
+
+  /// \brief Finds neghboring of |centerId| joint on feature id of |center|
+  /// which is contained in |edges| and fills |oneStepAside| with it.
+  /// \note If oneStepAside is empty no neghboring nodes were found.
+  /// \note Taking into account the way of setting restrictions almost always |oneStepAside|
+  /// will contain one or zero items. Meanwhile the it it's posible to draw map
+  /// the (wrong) way |oneStepAside| will contain any number of items.
+  void FindOneStepAsideRoadPoint(RoadPoint const & center, Joint::Id centerId,
+                                 vector<JointEdge> const & edges,
+                                 vector<Joint::Id> & oneStepAside) const;
+
+  bool GetIngoingAndOutgoingEdges(Joint::Id centerId, bool graphWithoutRestrictions,
+                                  vector<JointEdge> & ingoingEdges,
+                                  vector<JointEdge> & outgoingEdges);
+
+  bool ApplyRestrictionPrepareData(RestrictionPoint const & restrictionPoint,
+                                   RestrictionInfo & restrictionInfo);
+
+  void InsertToEdgeMapping(DirectedEdge const & key, DirectedEdge const & value);
 
   Geometry m_geometry;
   shared_ptr<EdgeEstimator> m_estimator;
   RoadIndex m_roadIndex;
   JointIndex m_jointIndex;
+
+  set<DirectedEdge> m_blockedEdges;
+  uint32_t m_nextFakeFeatureId = kStartFakeFeatureIds;
+  // Mapping from fake feature id to fake feature geometry.
+  map<uint32_t, RoadGeometry> m_fakeFeatureGeometry;
+  // Adding restrictions leads to disabling some edges and adding others.
+  // According to graph trasformation implemented in ApplyRestrictionNo() and
+  // ApplyRestrictionOnly() every |DirectedEdge| (a pair of joints) could:
+  // * disappears at all (in that case it's added to |m_blockedEdges|)
+  // * be transformed to another DirectedEdge. If so the mapping
+  //   is kept in |m_movedCrossings| and source edge is blocked (added to |m_blockedEdges|).
```

Comment:
|m_movedCrossings| in comment is left from an attempt to memorize whole restriction point instead of two edges. The comment will be fixed.